### PR TITLE
System class unification

### PR DESCRIPTION
**Pull request by vboggara-xilinx**
#### Problem solved by the commit
The system class boots up the system.  At core library load time, libxrt_core (shim library) instantiates concrete device factory class that are inserted into a singleton system class in xrt_coreutil.   When user host code creates an xrt::device object, then implementation goes through the singleton system class, find the proper device factory, and creates the device object.   The concrete device class is instantiated by the device factory at the shim level (libxrt_core).  Additional shim level objects are created through the interface of the device base class (xrt_core::device).  

#### How problem was solved, alternative solutions (if any) and why they were rejected
https://confluence.amd.com/pages/viewpage.action?spaceKey=AIE&title=XRT+ISHIM+Upgrade#XRTISHIMUpgrade-Systemclass
#### What has been tested and how, request additional testing if necessary
Validated below flows
PCIE - HW (testcase verification, plp_programming, xbutil and xbmgmt tools, all xbutil reports validation) , XMA
PCIE , EDGE - SW Emulation & HW Emulation
Windows - Compilation
#### Documentation impact (if any)
n/a

### CODE REVIEW

**Comment by vboggara-xilinx** [comment]
This PR has been created for Internal Review.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Review by stsoe — COMMENTED** [review]
This is a large number of changes, but looks manageable. 
It will take a while to review in detail.

File: src/runtime_src/core/common/dev_factory.h
Comment (stsoe) [inline src/runtime_src/core/common/dev_factory.h:12]: This is a device factory, so please class it `device_factory` and move the file to `device_factory.h`

File: src/runtime_src/core/common/dev_factory.h
Comment (stsoe) [inline src/runtime_src/core/common/dev_factory.h:19]: What's with the public data members, and does the concept of mgmt / user continue to be of use to XRT coreutil?  Can we maybe eliminate it?

File: src/runtime_src/core/common/dev_factory.h
Comment (stsoe) [inline src/runtime_src/core/common/dev_factory.h:34]: Where and how is this used?   Is the handle_type a shim handle?   We probably want to remove this API, which is why I ask where and how it is used.

File: src/runtime_src/core/common/dev_factory.h
Comment (stsoe) [inline src/runtime_src/core/common/dev_factory.h:36]: Be consistent in commenting, space after '//'
And do expand just a bit.

File: src/runtime_src/core/common/dev_factory.h
Comment (stsoe) [inline src/runtime_src/core/common/dev_factory.h:49]: `get_device_ready` and `get_device_total`  and comment what this means.

File: src/runtime_src/core/common/dev_factory.h
Comment (stsoe) [inline src/runtime_src/core/common/dev_factory.h:52]: What is this?

File: src/runtime_src/core/common/dev_factory.h
Comment (stsoe) [inline src/runtime_src/core/common/dev_factory.h:62]: I will need an explanation of the object ownership to understand this.   For example, I would think the shim library would create device factories upon load and pass ownership of these to XRT.   I am also questioning why this is not just a "register one factory" at a time interface?

**Review by stsoe — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/common/dev_factory.h
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/dev_factory.h:62]: yes soren, we are registering list of device_factory objects during loading of each driver( dlopen of .so) , we can make it one device registration instead of list.

File: src/runtime_src/core/common/dev_factory.h
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/dev_factory.h:52]: get_dev() is retruining devcie_factory object, its widely used in all of our code base like pcie/tools etc. we can rename it to proper one.

File: src/runtime_src/core/common/dev_factory.h
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/dev_factory.h:49]: sure

File: src/runtime_src/core/common/dev_factory.h
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/dev_factory.h:36]: sure

File: src/runtime_src/core/common/dev_factory.h
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/dev_factory.h:19]: we can move them to private and can access through public member functions, currently we are using  m_is_mgmt flag while creation of dev_factory class, like core/pcie/pcidev_linux.cpp has some functionality based on this.

File: src/runtime_src/core/common/dev_factory.h
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/dev_factory.h:12]: sure, will modify.

**Comment by vboggara-xilinx** [comment]
> This is a large number of changes, but looks manageable. It will take a while to review in detail.

yes soren, build pipeline also passing now.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by vboggara-xilinx** [comment]
please retest this.

**Comment by chvamshi-xilinx** [comment]
retest this please

File: src/runtime_src/core/common/device_factory.h
Comment (stsoe) [inline src/runtime_src/core/common/device_factory.h:11]: What does this mean?

File: src/runtime_src/core/common/device_factory.h
Comment (stsoe) [inline src/runtime_src/core/common/device_factory.h:19]: Leaving these bools public makes sense only if they can be `const` to prevent them from being changed by client code.  Anything else is bad design.  If `m_is_ready` cannot be initialized at construction time, then it cannot be a public data member.

Here is how to do is_mgmt (and do remove `m_`  for public data members).
```
  const bool is_mgmt {false};
  device_factory(bool isuser) 
    : m_is_msgmt{!isuser}
  {}
```
This prevents any modifications after construction.

File: src/runtime_src/core/common/device_factory.h
Comment (stsoe) [inline src/runtime_src/core/common/device_factory.h:21]:
```suggestion
  virtual ~device_factory() 
  {}
```

File: src/runtime_src/core/common/device_factory.h
Comment (stsoe) [inline src/runtime_src/core/common/device_factory.h:36]: Be consistent in commenting.  Yes, this one is quit obvious, but expand on what is returned, as it definitely isn't just BDF since it is a tuple of 4.  It's more like DBDF, but do explain.
```suggestion
  // Get BDF 
```

File: src/runtime_src/core/common/device_factory.h
Comment (stsoe) [inline src/runtime_src/core/common/device_factory.h:44]:
```suggestion
// Get list of device_factory objects which are in ready state from registered device list.
```

File: src/runtime_src/core/common/device_factory.h
Comment (stsoe) [inline src/runtime_src/core/common/device_factory.h:48]:
```suggestion
// Get total number of device_factory objects from registered device list.
```

File: src/runtime_src/core/common/device_factory.h
Comment (stsoe) [inline src/runtime_src/core/common/device_factory.h:52]:
```suggestion
// Get device_factory from list of registered device list.
```

File: src/runtime_src/core/common/device_factory.h
Comment (stsoe) [inline src/runtime_src/core/common/device_factory.h:54]:
```suggestion
get_device_factory(unsigned int index, bool user = true);
```

**Review by stsoe — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/common/device_factory.h
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/device_factory.h]: done

File: src/runtime_src/core/common/device_factory.h
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/device_factory.h]: done

File: src/runtime_src/core/common/device_factory.h
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/device_factory.h]: done

File: src/runtime_src/core/common/device_factory.h
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/device_factory.h]: done

File: src/runtime_src/core/common/device_factory.h
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/device_factory.h]: done

File: src/runtime_src/core/common/device_factory.h
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/device_factory.h]: made both mgmt and ready variables private.

**Review by stsoe — CHANGES_REQUESTED** [review]
@vboggara-xilinx It looks quite good.  I don't want to hold this up too long.  I have a bunch of nit picking issues, but we can ignore and I can fix when / if I have to edit the files at some point.  

My biggest concern right now is core/common/system.cpp which has too much conditional code and probably functions that do not belong at that level or should be hidden in platform specific files.  Not sure if my concern here makes any sense?

**Comment by vboggara-xilinx** [comment]
> @vboggara-xilinx It looks quite good. I don't want to hold this up too long. I have a bunch of nit picking issues, but we can ignore and I can fix when / if I have to edit the files at some point.
> 
> My biggest concern right now is core/common/system.cpp which has too much conditional code and probably functions that do not belong at that level or should be hidden in platform specific files. Not sure if my concern here makes any sense?
> 
> At least on my system (xsjsoren50.xilinx.com) the PR does not compile because of missing includes.
> 
> ```
> [ 62%] Linking CXX shared library libxilinxopencl.so
> Compiler wrapper: /tools/batonroot/rodin/devkits/lnx64/ccwrap/gccarwrap
> /proj/xsjhdstaff3/soeren/git/stsoe/XRT/src/runtime_src/core/pcie/linux/pcidrv_xocl.cpp: In constructor ‘{anonymous}::drv_xocl_reg::drv_xocl_reg()’:
> /proj/xsjhdstaff3/soeren/git/stsoe/XRT/src/runtime_src/core/pcie/linux/pcidrv_xocl.cpp:15:12: error: ‘cout’ is not a member of ‘std’
>    15 |       std::cout << "drv_xocl registration done" << std::endl;
>       |            ^~~~
> /proj/xsjhdstaff3/soeren/git/stsoe/XRT/src/runtime_src/core/pcie/linux/pcidrv_xocl.cpp:5:1: note: ‘std::cout’ is defined in header ‘<iostream>’; did you forget to ‘#include <iostream>’?
>     4 | #include "pcidrv_xocl.h"
>   +++ |+#include <iostream>
>     5 | 
> ```
> 
> After work-around / fix compile errors locally, I notice stray prints when running examples:
> 
> ```
> drv_xclmgmt registration done
> drv_xocl registration done
> system :: size of user_ready_list : 1
> system :: size of user_nonready_list : 0
> system :: size of mgmt_ready_list : 1
> system :: size of mgmt_nonready_list : 0
> ```
> 
> Please test this PR thoroughly before merging.

@stsoe , XRT build pipeline is passing and also I verified locally in all platforms, not seeing any compilation issue. Added those couts for testing purpose, will remove them before merging and will create separate system.cpp files for windows and linux.

File: src/runtime_src/core/common/system.cpp
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/system.cpp:54]: Sure soren, will fix it.

File: src/runtime_src/core/pcie/noop/pcidrv_noop.cxx
Comment (stsoe) [inline src/runtime_src/core/pcie/noop/pcidrv_noop.cxx:8]: Our coding convention says no indent within namespace.

File: src/runtime_src/core/pcie/noop/pcidev_noop.h
Comment (stsoe) [inline src/runtime_src/core/pcie/noop/pcidev_noop.h:13]: This is not a  pcidev, it is a device factory.  Naming is important and `pcidev` is just not good. Please call it what it is.
This would work
`class device_factory : public xrt_core::device_factory`
but maybe you'd prefer 
`class device_factory_noop : public device_factory`

File: src/runtime_src/core/pcie/noop/pcidev_noop.cxx
Comment (stsoe) [inline src/runtime_src/core/pcie/noop/pcidev_noop.cxx:6]: The namespace should be 
```suggestion
namespace xrt_core::noop {
```
Your are in the `noop` module, so keep types for `noop` in that namespace.

File: src/runtime_src/core/pcie/noop/pcidev_noop.cxx
Comment (stsoe) [inline src/runtime_src/core/pcie/noop/pcidev_noop.cxx:9]: Convention in my code is
```suggestion
  xrt_core::noop::
  create_device(device::handle_type handle, device::id_type id) const
```

File: src/runtime_src/core/pcie/noop/pcidev_noop.cxx
Comment (stsoe) [inline src/runtime_src/core/pcie/noop/pcidev_noop.cxx:16]:
```suggestion
  xrt_core::noop::
  create_shim(device::id_type id) const
```

File: src/runtime_src/core/pcie/noop/pcidev_noop.cxx
Comment (stsoe) [inline src/runtime_src/core/pcie/noop/pcidev_noop.cxx:18]: Too many spaces.

File: src/runtime_src/core/pcie/noop/pcidrv_noop.h
Comment (stsoe) [inline src/runtime_src/core/pcie/noop/pcidrv_noop.h:10]: wrong namespace

**Review by stsoe — CHANGES_REQUESTED** [review]
Please look at the batch of comments for `noop` directory and use that as playbook for all the other shims.   Naming is not correct, spacing is wrong also.

**Comment by chvamshi-xilinx** [comment]
Closing this as there is going to be another clean PR.

**Pull request closed without merging**
